### ISSUE TITLE
Spec: Gate 0 - Causal F0 Tracker Verification & Corpus Distribution Audit

### ISSUE THREAD

**Issue by HotoKokoa1204** [HotoKokoa1204/TimbreLite-Personalized-Ultra-Lightweight-Real-Time-Voice-Conversion#18]
## Problem Statement

During the EXP-1A full-corpus pitch extraction across the expanded 971-utterance corpus (481 Source + 490 Target), the Source Speaker (User) median fundamental frequency ($Q_{50}$) was calculated as $95.60\text{ Hz}$ ($IQR = 35.55\text{ Hz}$), and the Target Persona (Hu Tao) median was calculated as $329.60\text{ Hz}$ ($IQR = 151.39\text{ Hz}$). This implies a calculated pitch transposition of $+21.43\text{ semitones}$ ($3.45\times$ frequency multiplier).

However, this result introduces an immediate, critical empirical contradiction:
1. **Source Register Discrepancy**: Prior baseline observation on the canonical reference utterance `user_utt_0019` was approximately $131.5\text{ Hz}$. A measured corpus median of $95.60\text{ Hz}$ represents a massive drop (~36 Hz / 5.6 semitones).
2. **Target Register Discrepancy**: Prior baseline observation of Hu Tao target speech was approximately $393.4\text{ Hz}$, whereas the corpus median was measured at $329.60\text{ Hz}$.
3. **Core Risk**: Two competing hypotheses explain this disparity:
   - **Hypothesis A (Algorithmic Octave Halving / Tracking Artifact)**: The causal autocorrelation tracker ($W = 1280$, lookahead = 0) exhibits octave halving / subharmonic tracking on adult male low-pitched speech, locking onto $F0/2$ instead of the true fundamental frequency, and/or clipping high registers on female target speech.
   - **Hypothesis B (Genuine Corpus Distribution Shift)**: The addition of 192 newly curated clean utterances (expanding source corpus from 289 to 481 utterances) legitimately shifted the pooled speaker distribution, while individual utterances like `user_utt_0019` remain around ~131 Hz. Alternatively, frame-pooled statistics may over-weight sustained deep vowels compared to per-utterance median pooling.
4. **Downstream Danger**: Freezing $+21.43\text{ semitones}$ into the FiLM conditioning path without verifying tracker fidelity risks severe acoustic distortion (chipmunk formants from over-transposition, or deep male leakage from under-transposition).

Therefore, EXP-1A is marked as **Computed**, but **NOT Accepted**. Progression to EXP-1B (FiLM modulation and adapter ablation) is strictly blocked until Gate 0 is validated.

---

## Solution

Establish **Gate 0 (F0 Tracker Validation Gate)** to rigorously audit and validate the pitch tracking pipeline across the official **971-utterance corpus** (481 Source + 490 Target):
1. **Benchmark Utterance Re-evaluation**: Re-extract and audit the 3 canonical benchmark sentences (`user_utt_0019`, `user_utt_0130`, `user_utt_0190`) using the production `CausalF0Tracker`. Measure median, Q25, Q75, min, max, and voiced ratio, directly resolving whether `user_utt_0019` is tracked at ~131.5 Hz or ~83~95 Hz.
2. **Trajectory Continuity & Octave Jump Metric**: Implement and evaluate an adjacent-frame octave jump metric:
   $$r_t = \frac{F0_t}{F0_{t-1}} \quad \text{for continuous voiced frames } (VUV_t = 1, VUV_{t-1} = 1)$$
   Specifically measure the proportion of jumps falling into suspected octave-halving ($0.45 < r_t < 0.55$) and octave-doubling ($1.8 < r_t < 2.2$) intervals. Establish a strict acceptance threshold ($< 1.0\%$).
3. **Dual-Tracker Offline Reference Audit**: Employ a high-precision offline reference tracker (e.g., PyYIN / librosa / CREPE / Praat autocorrelation) as an audit oracle on identical waveforms to quantify Gross Pitch Error (GPE) and Pitch Error Centroids without polluting the real-time inference path.
4. **Target Persona High-Register Audit**: Verify whether Hu Tao's $329.60\text{ Hz}$ median suffers from octave halving on high-pitched vowel frames (400-800 Hz).
5. **Frame-Pooled vs. Utterance-Balanced Distribution Analysis**: Disentangle frame-duration bias by comparing frame-pooled quantiles against utterance-median-averaged quantiles across both the original 289 utterances and the new 192 utterances.
6. **Decoupled Production Configuration**: Isolate raw experimental logs (`outputs/exp1a_f0_stats.json`) from runtime deployment. Only after Gate 0 criteria are satisfied will the production configuration (`outputs/f0_production_config.json`) be minted and frozen for EXP-1B.

---

## User Stories

1. As a voice conversion researcher, I want to verify whether the causal pitch tracker suffers from octave halving on low adult male speech, so that I do not train neural conditioning on subharmonic pitch artifacts.
2. As a voice conversion researcher, I want to evaluate the 3 canonical test sentences (`user_utt_0019`, `user_utt_0130`, `user_utt_0190`) with the exact production tracker, so that I can determine if `user_utt_0019` is truly ~131 Hz or ~83~95 Hz.
3. As a voice conversion researcher, I want a quantitative octave-jump error metric ($0.45 < r_t < 0.55$ and $1.8 < r_t < 2.2$), so that I can objectively reject trackers with pitch-flip instabilities.
4. As an audio ML engineer, I want to compare the causal real-time tracker against a high-precision offline reference tracker (e.g. PyYIN/CREPE) during offline audit, so that I know exactly where causal approximations deviate from ground truth.
5. As an audio ML engineer, I want the offline reference tracker to remain strictly an audit tool, so that runtime inference maintains zero external pitch model dependencies and zero GPU contention.
6. As a voice conversion researcher, I want to audit Hu Tao's target distribution to confirm whether 329.60 Hz reflects true speech or high-frequency octave halving, so that the target register is accurately represented.
7. As an audio ML engineer, I want to formalize the official EXP-1A corpus scale as 971 utterances (481 source + 490 target), so that all subsequent research metrics reference consistent dataset boundaries.
8. As a voice conversion researcher, I want to compare frame-pooled statistics against per-utterance balanced statistics, so that long sustained vowels do not distort the global speaker pitch profile.
9. As a software developer, I want production F0 configuration (`f0_production_config.json`) to remain decoupled from raw experimental reports, so that unverified statistical shifts are never automatically committed into runtime configs.
10. As a software developer, I want Gate 0 to strictly block EXP-1B training, so that no compute or developer time is wasted training FiLM layers on questionable pitch statistics.
11. As a competitive gamer / end user, I want the converted voice to match Hu Tao's intended anime vocal register without sounding unnaturally high or chipmunk-like, so that role-play immersion is preserved.
12. As a competitive gamer / end user, I want converted speech to track pitch inflections smoothly without abrupt octave jumps or clicks, so that communication in Discord sounds natural and intelligible.
13. As a dataset curator, I want the 192 newly cleaned source utterances to be preserved, so that the expanded voice volume can be utilized once tracker validity is proven.
14. As an audio QA engineer, I want continuous F0 trajectory plots for benchmark sentences rendered visually, so that manual inspection can instantly spot pitch drops and edge artifacts.
15. As a systems architect, I want Gate 0 verification to run fully automated in CI/CLI, so that any future changes to tracker parameters (hop size, window, thresholds) can be instantly regression-tested.

---

## Implementation Decisions

1. **Gate 0 Verification Architecture**:
   - Establish an independent verification harness that isolates acoustic tracking validation from downstream neural model weights.
   - The harness processes audio files and produces structured validation metrics and diagnostic reports without touching runtime adapter code.

2. **Octave Jump Formalization**:
   - For consecutive voiced frames, calculate the frequency ratio $r_t = F0_t / F0_{t-1}$.
   - Classify transitions where $r_t \in [0.45, 0.55]$ as octave-halving candidates.
   - Classify transitions where $r_t \in [1.80, 2.20]$ as octave-doubling candidates.
   - Total octave error candidate rate:
     $$E_{oct} = \frac{N_{\text{halving}} + N_{\text{doubling}}}{N_{\text{voiced transitions}}}$$

3. **Dual-Tracker Validation Protocol**:
   - Use a high-precision offline reference (e.g. PyYIN or Praat-style normalized autocorrelation with full non-causal lookahead) to establish pitch ground truth on reference speech.
   - Calculate Gross Pitch Error (GPE) as the percentage of voiced frames where causal F0 deviates from offline reference F0 by more than $20\%$.
   - Calculate Mean Absolute Error in Cents on consensus voiced frames ($GPE < 20\%$).

4. **Statistical Aggregation Modes**:
   - Implement dual pooling comparisons:
     - **Frame-Pooled**: Concatenate all voiced frames across the corpus and compute percentiles across all frames.
     - **Utterance-Balanced**: Compute median and IQR per utterance, then compute the macro distribution across utterances.
   - Compare the original 289 utterances against the expanded 481 utterances to isolate dataset expansion effects from algorithmic tracking bias.

5. **Decoupled Production Configuration Schema**:
   - Retain `exp1a_f0_stats.json` purely as diagnostic output.
   - Only after Gate 0 verification passes will `f0_production_config.json` be generated with the approved schema:
```json
{
  "source": {
    "median_hz": 0.0,
    "iqr_hz": 0.0,
    "median_log_f0": 0.0,
    "iqr_log_f0": 0.0
  },
  "target": {
    "median_hz": 0.0,
    "iqr_hz": 0.0,
    "median_log_f0": 0.0,
    "iqr_log_f0": 0.0
  },
  "scale_log_iqr": 1.0,
  "pitch_bias_semitones": 0.0,
  "gate_verified": true
}
```

---

## Testing Decisions

1. **Primary Testing Seam**:
   - The **F0 Evaluation Seam**: Waveform input to statistical report output (`AudioWaveform -> F0TrackerValidationHarness -> ValidationReport`).
   - This seam is the highest possible boundary for pitch extraction, completely decoupled from neural network weights or downstream audio device I/O.
2. **Definition of a Good Test**:
   - Tests verify observable physical acoustic properties (e.g., tracking a known harmonic voice within $\pm 3\text{ Hz}$, maintaining $E_{oct} < 1.0\%$, and achieving consensus with offline reference), never internal private variables or circular buffer pointer indices.
3. **Modules Tested**:
   - Causal F0 Tracker, Octave Jump Detector, Reference Tracker Comparator, and Production Config Exporter.
4. **Prior Art**:
   - `tests/test_f0.py` (causal tracking on synthetic harmonic tones and frequency bounds).
   - `scripts/run_exp1a_f0_audit.py` (corpus-wide batch extraction).

---

## Out of Scope

- Training FiLM F0 modulation layers (EXP-1B).
- Modifying neural adapter architecture or retraining checkpoint weights.
- Waveform-level pitch shifting (e.g., phase vocoder or WSOLA pre-processing).
- Retraining or introducing neural F0 prediction models.
- Altering the approved causal window size (1280 samples / 53.33ms) or algorithmic lookahead (strictly 0).

---

## Further Notes

- Formal corpus scale is established as **971 utterances** (481 Source = 431 train + 50 val; 490 Target = 441 train + 49 val).
- Offline reference trackers are strictly restricted to development auditing and validation; live streaming inference remains pure CPU causal autocorrelation with zero external pitch dependencies.

**Comment by HotoKokoa1204** [comment]
Part of #8

**Comment by HotoKokoa1204** [comment]
### Gate 0 Verification: PASSED (Signed Off)

All 8 Gate 0 criteria and 4 Hard Requirements have been verified and passed:
- **G0-1**: Source Train {oct} = 0.57\% < 1.0\%$; Target Train {oct} = 0.15\% < 1.0\%$.
- **G0-2**: Target Train High-Register Halving Rate ($>400\text{ Hz}$) $= 0.94\% < 1.0\%$.
- **G0-3**: Benchmark sentences PyIN consensual Oracle GPE $= 8.67\% < 10.0\%$.
- **G0-4 / G0-5**: Benchmark fixtures (\user_utt_0019\, \user_utt_0130\, \user_utt_0190\) track within target bounds and reject 2nd harmonics ($>65\text{ Hz}$ separation). \user_utt_0190\ tagged \F0_REGRESSION_WARNING\ for monitoring.
- **G0-6**: Bitwise streaming parity between \process_chunk\ and \process_utterance\ verified (exact 0.0 diff).
- **G0-7**: Log-domain IQR quantiles ($\text{IQR}_{\log F0} = Q_{75}(\ln F0) - Q_{25}(\ln F0)$) mathematically verified.
- **G0-8**: Full pytest test suite passes (78 passed, 0 failed).

Decoupled production configuration minted: \outputs/f0_production_config.json\ (Source Median: 91.62 Hz, Target Median: 326.12 Hz, Span: +21.98 st, Scale: 1.4580).
Commits: \c4f0b63\, \e4d79a5\.
Closing as completed to unblock EXP-1B.
